File: src/generate.py
import os
import sys
import random
import requests
from io import BytesIO
from PIL import Image, ImageDraw, ImageFont
from datetime import datetime

# Configuration
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.dirname(SCRIPT_DIR)
FONT_PATH = os.path.join(PROJECT_ROOT, "assets", "consolasb.ttf")
OUTPUT_PATH = os.path.join(PROJECT_ROOT, "final_status.jpg")
OVERLAY_IMAGE_PATH = os.path.join(PROJECT_ROOT, "assets", "overlay_image.png")
FALLBACK_IMAGE = os.path.join(PROJECT_ROOT, "assets", "img_base.png")

# Dimensions & Output
FINAL_SIZE = (720, 1278)
TEXT_POSITION = (360, 700)  # Centered horizontally
PROGRESS_BAR_START = (81, 860)
PROGRESS_BAR_MAX_WIDTH = 558
PROGRESS_BAR_HEIGHT = 30
FONT_SIZE = 75

# Colors
WHITE = (255, 255, 255)
BAR_COLOR = (255, 255, 255)

def calculate_year_progress():
    """Calculate the current year's progress as a percentage."""
    now = datetime.now()
    year_start = datetime(now.year, 1, 1)
    year_end = datetime(now.year + 1, 1, 1)
    total_seconds = (year_end - year_start).total_seconds()
    elapsed_seconds = (now - year_start).total_seconds()
    return round((elapsed_seconds / total_seconds) * 100, 2)

def get_background():
    """Fetch random landscape from picsum or fallback to local template."""
    try:
        response = requests.get("https://picsum.photos/1920/1080", timeout=10)
        response.raise_for_status()
        return Image.open(BytesIO(response.content)).convert("RGBA")
    except Exception as e:
        print(f"Warning: Failed to fetch remote background: {e}. Falling back.")
        if os.path.exists(FALLBACK_IMAGE):
            return Image.open(FALLBACK_IMAGE).convert("RGBA")
        return Image.new("RGBA", FINAL_SIZE, (0, 0, 0, 255))

def process_background(bg_img):
    """Aspect-fill resize and center-crop to 720x1278 (no stretching)."""
    target_w, target_h = FINAL_SIZE
    img_w, img_h = bg_img.size
    
    # Calculate scale factor for aspect fill (smallest dimension fits)
    scale = max(target_w / img_w, target_h / img_h)
    new_w = int(img_w * scale)
    new_h = int(img_h * scale)
    
    bg_img = bg_img.resize((new_w, new_h), Image.Resampling.LANCZOS)
    
    # Center crop to target size
    left = (new_w - target_w) / 2
    top = (new_h - target_h) / 2
    right = left + target_w
    bottom = top + target_h
    
    return bg_img.crop((left, top, right, bottom))

def get_random_texture():
    """Pick a random texture from assets/textures/."""
    texture_dir = os.path.join(PROJECT_ROOT, "assets", "textures")
    textures = [f for f in os.listdir(texture_dir) if f.lower().endswith(('.png', '.jpg'))]
    if not textures:
        return None
    texture_path = os.path.join(texture_dir, random.choice(textures))
    return Image.open(texture_path).convert("RGBA")

def generate_status_image():
    percentage = calculate_year_progress()
    bg_img = get_background()
    bg_img = process_background(bg_img)
    
    # 30% Opaque black overlay for readability
    dark_overlay = Image.new("RGBA", bg_img.size, (0, 0, 0, int(255 * 0.3)))
    bg_img = Image.alpha_composite(bg_img, dark_overlay)
    
    # Texture layer
    texture_img = get_random_texture()
    if texture_img:
        texture_img = texture_img.resize(FINAL_SIZE, Image.Resampling.LANCZOS)
        bg_img = Image.alpha_composite(bg_img, texture_img)
    
    # Composite UI Overlay if it exists
    if os.path.exists(OVERLAY_IMAGE_PATH):
        ui_overlay = Image.open(OVERLAY_IMAGE_PATH).convert("RGBA")
        ui_overlay = ui_overlay.resize(FINAL_SIZE, Image.Resampling.LANCZOS)
        bg_img = Image.alpha_composite(bg_img, ui_overlay)
    else:
        print(f"Warning: {OVERLAY_IMAGE_PATH} not found. Skipping UI overlay.")
    
    # Prepare to draw text/bar
    bg_img = bg_img.convert("RGB")
    draw = ImageDraw.Draw(bg_img)
    
    # Load Font
    try:
        font = ImageFont.truetype(FONT_PATH, FONT_SIZE)
    except IOError:
        print(f"Warning: {FONT_PATH} missing. Using default.")
        font = ImageFont.load_default()
        
    # Draw Text
    text = f"{percentage:.2f}%"
    draw.text(TEXT_POSITION, text, fill=WHITE, font=font, anchor="mm")
    
    # Draw Progress Bar
    fill_width = int((percentage / 100) * PROGRESS_BAR_MAX_WIDTH)
    x1, y1 = PROGRESS_BAR_START
    x2 = x1 + fill_width
    y2 = y1 + PROGRESS_BAR_HEIGHT
    draw.rectangle([x1, y1, x2, y2], fill=BAR_COLOR)
    
    # Save Output
    bg_img.save(OUTPUT_PATH, "JPEG", quality=95)
    print(f"Generated: {OUTPUT_PATH}")
    print(f"Year Progress: {percentage:.2f}%")

if __name__ == "__main__":
    generate_status_image()

File: src/generate.js
const sharp = require('sharp');
const { createCanvas } = require('canvas');
const fs = require('fs');
const path = require('path');

const PROJECT_ROOT = path.dirname(__dirname);
const OUTPUT_PATH = path.join(PROJECT_ROOT, 'final_status.jpg');
const OVERLAY_IMAGE_PATH = path.join(PROJECT_ROOT, 'assets', 'overlay_image.png');
const FALLBACK_IMAGE = path.join(PROJECT_ROOT, 'assets', 'img_base.png');
const TEXTURE_DIR = path.join(PROJECT_ROOT, 'assets', 'textures');

const FINAL_WIDTH = 720;
const FINAL_HEIGHT = 1278;
const TEXT_X = 360;
const TEXT_Y = 700;
const BAR_X = 81;
const BAR_Y = 860;
const BAR_MAX_WIDTH = 558;
const BAR_HEIGHT = 30;
const FONT_SIZE = 75;

function calculateYearProgress() {
    const now = new Date();
    const yearStart = new Date(now.getFullYear(), 0, 1);
    const yearEnd = new Date(now.getFullYear() + 1, 0, 1);
    const totalMs = yearEnd - yearStart;
    const elapsedMs = now - yearStart;
    return Math.round((elapsedMs / totalMs) * 10000) / 100;
}

async function fetchBackground() {
    try {
        const res = await fetch('https://picsum.photos/1920/1080', { redirect: 'follow' });
        if (!res.ok) throw new Error(`HTTP ${res.status}`);
        const buf = Buffer.from(await res.arrayBuffer());
        console.log(`Fetched remote background (${buf.length} bytes)`);
        return buf;
    } catch (e) {
        console.log(`Warning: Failed to fetch remote background: ${e.message}. Falling back.`);
        if (fs.existsSync(FALLBACK_IMAGE)) {
            const buf = fs.readFileSync(FALLBACK_IMAGE);
            console.log(`Using fallback image (${buf.length} bytes)`);
            return buf;
        }
        console.log('No fallback image found, using black background.');
        return null;
    }
}

function getRandomTexture() {
    if (!fs.existsSync(TEXTURE_DIR)) return null;
    const files = fs.readdirSync(TEXTURE_DIR).filter(f => /\.(png|jpg|jpeg)$/i.test(f));
    if (files.length === 0) return null;
    const pick = files[Math.floor(Math.random() * files.length)];
    return path.join(TEXTURE_DIR, pick);
}

function makeDarkOverlay(width, height, opacity) {
    const channels = 4;
    const buf = Buffer.alloc(width * height * channels);
    const alpha = Math.round(255 * opacity);
    for (let i = 0; i < buf.length; i += channels) {
        buf[i] = 0;     // R
        buf[i + 1] = 0; // G
        buf[i + 2] = 0; // B
        buf[i + 3] = alpha; // A
    }
    return sharp(buf, { raw: { width, height, channels } }).png().toBuffer();
}

async function generateStatusImage() {
    const percentage = calculateYearProgress();
    console.log(`Year Progress: ${percentage.toFixed(2)}%`);

    // 1. Start with background
    const bgBuf = await fetchBackground();
    let base;
    if (bgBuf) {
        base = sharp(bgBuf).resize(FINAL_WIDTH, FINAL_HEIGHT, { fit: 'cover', position: 'centre' });
    } else {
        base = sharp({
            create: {
                width: FINAL_WIDTH,
                height: FINAL_HEIGHT,
                channels: 4,
                background: { r: 0, g: 0, b: 0, alpha: 255 }
            }
        });
    }

    // 2. Dark overlay (30% opacity)
    const darkOverlay = await makeDarkOverlay(FINAL_WIDTH, FINAL_HEIGHT, 0.3);

    const composites = [{ input: darkOverlay }];

    // 3. Random texture (already has proper alpha)
    const texturePath = getRandomTexture();
    if (texturePath) {
        const textureBuf = await sharp(texturePath)
            .resize(FINAL_WIDTH, FINAL_HEIGHT, { fit: 'fill' })
            .png().toBuffer();
        composites.push({ input: textureBuf });
    }

    // 4. UI overlay (use as-is, like Python did)
    if (fs.existsSync(OVERLAY_IMAGE_PATH)) {
        const overlayBuf = await sharp(OVERLAY_IMAGE_PATH)
            .resize(FINAL_WIDTH, FINAL_HEIGHT, { fit: 'fill' })
            .png().toBuffer();
        composites.push({ input: overlayBuf });
    } else {
        console.log(`Warning: ${OVERLAY_IMAGE_PATH} not found. Skipping UI overlay.`);
    }

    // 5. Draw text + progress bar on canvas
    const canvas = createCanvas(FINAL_WIDTH, FINAL_HEIGHT);
    const ctx = canvas.getContext('2d');

    ctx.fillStyle = '#ffffff';
    ctx.font = `bold ${FONT_SIZE}px Consolas`;
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    ctx.fillText(`${percentage.toFixed(2)}%`, TEXT_X, TEXT_Y);

    const fillWidth = Math.round((percentage / 100) * BAR_MAX_WIDTH);
    ctx.fillStyle = '#ffffff';
    ctx.fillRect(BAR_X, BAR_Y, fillWidth, BAR_HEIGHT);

    composites.push({ input: canvas.toBuffer('image/png') });

    // 6. Composite all layers and save
    await base
        .composite(composites)
        .jpeg({ quality: 95 })
        .toFile(OUTPUT_PATH);

    console.log(`Generated: ${OUTPUT_PATH}`);
}

generateStatusImage().catch(err => {
    console.error('Failed to generate image:', err);
    process.exit(1);
});
